src/Microsoft.TestPlatform.Utilities/CodeCoverageRunSettingsProcessor.cs

```diff
@@ -0,0 +1,484 @@
+            {
+                new Exclusion(new List<string> { "ModulePaths", "Exclude" }),
+                new Exclusion(new List<string> { "Attributes", "Exclude" }),
+                new Exclusion(new List<string> { "Sources", "Exclude" }),
+                new Exclusion(new List<string> { "Functions", "Exclude" })
+            };
+        }
+        #endregion
+
+        #region Public Interface
+        /// <summary>
+        /// Processes the current settings for the code coverage data collector.
+        /// </summary>
+        /// 
+        /// <param name="currentSettings">The code coverage settings.</param>
+        /// 
+        /// <returns>An updated version of the current run settings.</returns>
+        public XmlNode Process(string currentSettings)
+        {
+            if (string.IsNullOrEmpty(currentSettings))
+            {
+                return null;
+            }
+
+            // Load current settings from string.
+            var document = new XmlDocument();
+            document.LoadXml(currentSettings);
+
+            return this.Process(document.DocumentElement);
+        }
+
+        /// <summary>
+        /// Processes the current settings for the code coverage data collector.
+        /// </summary>
+        /// 
+        /// <param name="currentSettingsDocument">
+        /// The code coverage settings document.
+        /// </param>
+        /// 
+        /// <returns>An updated version of the current run settings.</returns>
+        public XmlNode Process(XmlDocument currentSettingsDocument)
+        {
+            if (currentSettingsDocument == null)
+            {
+                return null;
+            }
+
+            return this.Process(currentSettingsDocument.DocumentElement);
+        }
+
+        /// <summary>
+        /// Processes the current settings for the code coverage data collector.
+        /// </summary>
+        /// 
+        /// <param name="currentSettingsRootNode">The code coverage root element.</param>
+        /// 
+        /// <returns>An updated version of the current run settings.</returns>
+        public XmlNode Process(XmlNode currentSettingsRootNode)
+        {
+            if (currentSettingsRootNode == null)
+            {
+                return null;
+            }
+
+            // Get the code coverage node from the current settings. If unable to get any
+            // particular component down the path just add the default values for that component
+            // from the default settings document and return since there's nothing else to be done.
+            var codeCoveragePathComponents = new List<string>() { "CodeCoverage" };
+            var currentCodeCoverageNode = this.SelectNodeOrAddDefaults(
+                currentSettingsRootNode,
+                this.defaultSettingsRootNode,
+                codeCoveragePathComponents);
+
+            if (currentCodeCoverageNode == null)
+            {
+                return currentSettingsRootNode;
```

Comment:
this is some error case, right? i think return null; is more readable. Please add some comment

Reply:
Yes, it's indeed an error case but we don't return `currentCodeCoverageNode` which is null, but `currentSettingsRootNode` which is not null here.

The effect of the call above, `this.SelectNodeOrAddDefaults(...)`, is that it either correctly extracts the code coverage node and returns it, or it adds defaults for it and returns null. In case null is returned we bail out by returning the root node which is not null.

You're correct that the reason for bailing out is not apparent so I added a comment there.